File: lintcode/python/1318_contains_duplicate_III.py
#!/usr/bin/python -t

# hash table

class Solution:
    """
    @param nums: the given array
    @param k: the given k
    @param t: the given t
    @return: whether there are two distinct indices i and j in the array such that the absolute difference between nums[i] and nums[j] is at most t and the absolute difference between i and j is at most k.
    """
    def containsNearbyAlmostDuplicate(self, nums, k, t):
        # Write your code here
        if t == 0 and len(nums) == len(set(nums)):
            return False
        for i in range(len(nums)-1):
            for j in range(i+1, min(i+1+k, len(nums))):
                if abs(nums[i]-nums[j]) <= t:
                    return True 
        return False


class Solution:
    """
    @param nums: the given array
    @param k: the given k
    @param t: the given t
    @return: whether there are two distinct indices i and j in the array such that the absolute difference between nums[i] and nums[j] is at most t and the absolute difference between i and j is at most k.
    """
    def containsNearbyAlmostDuplicate(self, nums, k, t):
        # Write your code here

        buckets = {}
        for i, v in enumerate(nums):
            # t == 0 is a special case where we only have to check the bucket
            # that v is in.
            bucketNum, offset = (v / t, 1) if t else (v, 0)
            for idx in xrange(bucketNum - offset, bucketNum + offset + 1):
                if idx in buckets and abs(buckets[idx] - nums[i]) <= t:
                    return True

            buckets[bucketNum] = nums[i]
            if len(buckets) > k:
                # Remove the bucket which is too far away. Beware of zero t.
                del buckets[nums[i - k] / t if t else nums[i - k]]

        return False

File: lintcode/golang/1318_contains_duplicate_III.go

// hash

/**
 * @param nums: the given array
 * @param k: the given k
 * @param t: the given t
 * @return: whether there are two distinct indices i and j in the array such that the absolute difference between nums[i] and nums[j] is at most t and the absolute difference between i and j is at most k.
 */
func containsNearbyAlmostDuplicate (nums []int, k int, t int) bool {
    // Write your code here
    n := len(nums)
    if n == 0 {
        return false
    }

    bucket := make(map[int]int)

    for i, v := range nums {
        bucketNum := v
        offset := 0
        if t != 0 {
            bucketNum = v/t
            offset = 1
        }

        for idx := bucketNum-offset; idx < bucketNum+offset+1; idx++ {
            if val, ok := bucket[idx]; ok {
                if abs(val - v) <= t {
                    return true
                }
            }
        }
        bucket[bucketNum] = v
        if len(bucket) > k {
            if t == 0 {
                delete(bucket, nums[i-k])
            } else {
                delete(bucket, nums[i-k]/t)
            }
        }
    }

    return false
}

func abs(a int) int {
    if a < 0 {
        return -a
    }
    return a
}

